Feat/collab service

zechajw opened the pull request:
- [X] Add initial layout for collaboration page in frontend under /room route
- [X] Add sockets and other APIs necessary for collaboration service
- [X] Expose zustand useRoomStore for integration with matching-service
- [X] Add different language support to dropdown
- [X] Add /:roomId parameter to room route
- [X] Add /constants/languages.ts for language support
- [X] Add pulumi, github, AWS integration

zechajw committed:
Add collab-service frontend

zechajw committed:
Add: Collab-service backend

zechajw committed:
Merge branch 'staging' into feat/collab-service

raynerljm commented on backend/collab-service/.gitignore:
```diff
@@ -0,0 +1,130 @@
+.yarn-integrity
+
+# dotenv environment variable files
+.env
```
can add `.env.development` also

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+
+const port = process.env.port || 8000;
+
+const server = http.createServer(app);
```
just a question for my own learning - what do this do? compared to just using the express app as a server directly

zechajw replied:
express.app uses http server under the hood -> but the reason why i created it this way is so that i could attach the socket.io server to the same http server instance

another way to do this would be to create the socket server using app.listen() which returns the underlying http server, but I did it this way for better clarity

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+import { ChangeSet, Text } from "@codemirror/state"
+import { getDocumentInfo, getPullUpdatesInfo, getUpdateInfo, updateDocInfo } from "./models/rooms.model"
+
+const port = process.env.port || 8000;
```
can make it default to ur service's assigned port (i.e. 8005)

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+const io = new Server(server, {
+    // options
+    cors: {
+        origin: "*",
```
can refer to the node env for the frontend domain

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+})
+
+export const redisClient = createClient()
```
so i'm assuming to run this, you will need to run redis locally on ur system? is it possible to either 
1. use docker-compose to run a redis client for u and create the redis client accordingly (e.g. depending on de, staging, prod)
2. or just use an in-memory redis client so it runs entirely within the server

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+
+io.on('connection', (socket) => {
+
+    socket.on('pullUpdates', (version: number, roomId: string) => {
```
possible to abstract out `pullUpdates` to an enum

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+        } else {
+            let { updates, pending, doc } = updateInfo.data
+            try {
+                if (version != updates.length) {
```
```suggestion
                if (version !== updates.length) {
```

strict equality

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+                } else {
+                    let newUpdates = [];
+                    for (let update of docUpdates) {
+                        let changes = ChangeSet.fromJSON(update.changes);
```
doesn't seem like `changes` is being mutated so can declare it as a const

raynerljm commented on backend/collab-service/src/server.ts:
```diff
@@ -0,0 +1,93 @@
+import http from "http"
```
this file has a lot specific logic so would be helpful to add comments explaining what each section does

raynerljm commented on backend/collab-service/src/routes/room.ts:
```diff
@@ -0,0 +1,62 @@
+const roomRouter = express.Router()
+
+roomRouter.use((req: Request, res: Response, next: NextFunction) => {
+    console.log('Room Request received')
```
nice i like this

raynerljm commented on backend/collab-service/src/routes/room.ts:
```diff
@@ -0,0 +1,62 @@
+        const createRoomResponse = createRoom(roomId)
+
+        if (!(createRoomResponse.data)) {
+            return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json({ message: createRoomResponse.error ?? "" })
```
is it possible to follow the `jsend` format of `{ status, data }`

raynerljm commented on backend/collab-service/src/models/rooms.model.ts:
```diff
@@ -0,0 +1,123 @@
+}
+
+export const createRoom = (roomId: string): ModelResponse<roomsType.createRoomData> => {
+    if (roomId in rooms) {
```
oh wow i never knew u could do this

raynerljm commented on backend/collab-service/src/models/rooms.model.ts:
```diff
@@ -0,0 +1,123 @@
+import { Update } from "@codemirror/collab"
+import * as roomsType from "../types/rooms.type";
+
+export const rooms: Record<string, roomsType.roomInfo> = {
```
im a bit confused, so if u r using this object here what is the redis client for?

zechajw replied:
the redis client not used. will remove

raynerljm commented on frontend/app/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,25 @@
+interface Option {
+    id: number;
+    description: string;
+}
```
as this type could be possibly used in other components, can you move this to `components/Dropdown/Dropdown.types.ts` and export it from there

and in `components/Dropdown/index.ts` you can do `export * from "./Dropdown.types.ts"` also

raynerljm commented on frontend/app/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -0,0 +1,25 @@
+            <Text textStyle="text-sm">{title}</Text>
+            <Select placeholder={placeholder ?? "Select Option"} size="xs" textColor="white">
+                {options.map(({ id, description }) =>
+                    <option key="id" value={id}>{description}</option>
```
```suggestion
                    <option key={id} value={id}>{description}</option>
```

did u mean to do this

raynerljm commented on frontend/app/src/features/room/components/info/DifficultyBadge.tsx:
```diff
@@ -0,0 +1,19 @@
+
+export const DifficultyBadge: React.FC<Props> = ({ difficulty }) => {
+
+    const colorSchemes: Record<DifficultyType, string> = {
```
up to u but this could be refactored to something in `/theme` because I foresee this color scheme being used by the select preferences card component as well

raynerljm replied:
either that or move the `DifficultyBadge` into `/components` for more general use

you can also shift the `colorSchemes` object outside the component so it doesn't get re-defined on every rerender

raynerljm commented on frontend/app/src/features/room/components/info/DifficultyBadge.tsx:
```diff
@@ -0,0 +1,19 @@
+        "Hard": "red"
+    }
+
+    return <Badge colorScheme={colorSchemes[difficulty]} variant="outline">Easy</Badge>
```
```suggestion
    return <Badge colorScheme={colorSchemes[difficulty]} variant="outline">{difficulty}</Badge>
```

did u mean to do this?

zechajw replied:
yup LOL i forgot to change that part

raynerljm commented on frontend/app/src/features/room/components/info/DifficultyBadge.tsx:
```diff
@@ -0,0 +1,19 @@
+import { DifficultyType } from "@/constants/question"
```
i like that u abstracted this into its own component - good thinking

raynerljm commented on frontend/app/src/features/room/components/info/InfoBar.tsx:
```diff
@@ -0,0 +1,24 @@
+import { DifficultyType } from "@/constants/question"
+import { UserCard } from "@/features/matching"
+
+interface Props {
```
nit: for all my previous components i named the props according to the component name (e.g. `InfoBarProps`) might be good to follow this standard across ur components

zechajw replied:
Ok Ill change all of this

joelwongjy replied:
I actually liked it as Props cuz its just used internally and once only HAHA

raynerljm commented on frontend/app/src/features/room/components/info/InfoBar.tsx:
```diff
@@ -0,0 +1,24 @@
+    topic: string;
+}
+
+export const InfoBar: React.FC<Props> = ({ difficulty, topic }) => {
```
nit: if you refer to other components, I didn't use `FC` and just directly typed the input as the props - I would recommend following this convention too so the codebase is standardize (neither way is more correct or wrong afaik)

raynerljm commented on frontend/app/src/features/room/components/info/InfoBar.tsx:
```diff
@@ -0,0 +1,24 @@
+export const InfoBar: React.FC<Props> = ({ difficulty, topic }) => {
+    return (
+        <HStack justifyContent={"space-between"}>
+            <HStack>
```
i haven't actually seen this in local but I foresee that u need to add some form of `gap`

raynerljm commented on frontend/app/src/features/room/components/code/QuestionDetails.tsx:
```diff
@@ -0,0 +1,13 @@
+
+export const QuestionDetails: React.FC<Props> = ({ details }) => {
+    return (
+        <Box flex={1} border="1px" borderColor={"ActiveBorder"} rounded="md" height="full">
```
haven't seen this in local yet but why not use the `Card` component and override the border colour? would like to hear ur thoughts

raynerljm commented on frontend/app/src/pages/room/index.tsx:
```diff
@@ -0,0 +1,16 @@
+function RoomPage() {
+    return (
+        <Page display="grid" placeItems="center">
+            <VStack marginBottom="4" flexDirection="column" align="left" height="full" width="full">
```
good to add units for the margin bottom to be clearer

raynerljm commented on frontend/app/src/pages/room/index.tsx:
```diff
@@ -0,0 +1,16 @@
+    return (
+        <Page display="grid" placeItems="center">
+            <VStack marginBottom="4" flexDirection="column" align="left" height="full" width="full">
+                <InfoBar difficulty={"Easy"} topic={"Data Structures & Algorithms"} />
```
can add a `//TODO:` since these hard coded values should eventually be changed

raynerljm commented on frontend/app/src/features/room/components/code/CodeEditor.tsx:
```diff
@@ -0,0 +1,65 @@
+
+    const [doc, setDoc] = useState<string | null>(null)
+    const [version, setVersion] = useState<number | null>(null)
+    const [roomId, setRoomId] = useState<string>("1")
```
don't have to define useState type if it can be inferred

raynerljm commented on frontend/app/src/features/room/components/code/CodeEditor.tsx:
```diff
@@ -0,0 +1,65 @@
+    const [version, setVersion] = useState<number | null>(null)
+    const [roomId, setRoomId] = useState<string>("1")
+
+    useEffect(() => {
```
just an idea - i understand that there's a lot more unique logic here with websockets so it's not very straightforward to integrate with react-query (which by right should be the way we sync client and server data)

but if u want to pursue it it could be an interesting learning experience
https://tkdodo.eu/blog/using-web-sockets-with-react-query

joelwongjy replied:
cool idea i will look into it also

raynerljm commented on frontend/app/src/features/room/components/code/CodeEditor.tsx:
```diff
@@ -0,0 +1,65 @@
+
+
+
+    if (version != null && doc != null) {
```
try to use strict equality
```suggestion
    if (version !== null && doc !== null) {
```

raynerljm commented on frontend/app/src/features/room/components/code/CodeEditor.tsx:
```diff
@@ -0,0 +1,65 @@
+                extensions={[
+                    indentUnit.of("\t"),
+                    basicSetup(),
+                    langs.python(),
```
possible to abstract the supported langs into a constant somewhere?

e.g. 
`/room/constants/codeEditor.ts`

raynerljm commented on frontend/app/src/features/room/components/code/CodeEditor.tsx:
```diff
@@ -0,0 +1,65 @@
+            />
+        </Box>)
+    } else {
+        return <Box>Loading...</Box>
```
can use chakra UI loading spinner?

might need to think about the logic as well because if the finding of the room fails, usess will see loading forever

raynerljm commented on frontend/app/src/features/room/components/code/Collaborator.tsx:
```diff
@@ -0,0 +1,65 @@
+    const [socket, setSocket] = useState<Socket | null>(null)
+
+    useEffect(() => {
+        const connectSocket = io("http://localhost:8000")
```
can refer to this as an env var

raynerljm replied:
we will eventually want to try referring to this thru the API gateway

raynerljm commented on frontend/app/src/features/room/components/code/Collaborator.tsx:
```diff
@@ -0,0 +1,65 @@
+
+    const options = (
+        <HStack>
+            <Dropdown title="Question" options={[]} />
```
can add `//TODO:` to update the options once they are in

raynerljm commented on frontend/app/src/features/room/components/code/Collaborator.tsx:
```diff
@@ -0,0 +1,65 @@
+    )
+
+    const codeEditor = (
+        socket ? <CodeEditor socket={socket} /> : <Box>Loading</Box>
```
can replace the loading with loading spinner

raynerljm commented on frontend/app/src/features/room/components/code/Collaborator.tsx:
```diff
@@ -0,0 +1,65 @@
+                {renderQuestion ? visibleView : hiddenView}
+            </Box>
+            <Box>
+                <Button size="xs" background="#3F3F46" textColor="white" onClick={() => setRenderQuestion(!renderQuestion)}>{renderQuestion ? "Hide" : "Show"} Question</Button>
```
can refer to `frontend/app/src/theme/foundations/colors.ts`
- the `#3F3F46` is already defined as `light.500`
- so you can just define `<Button colorScheme="light" ...`
- this also automatically turns ur textColor to white as it auto detects the contrast

raynerljm reviewed (changes requested):
mostly looks good! left some clarifying comments about convention, clarity and structure
- pulumi stuff is fixed so can add the pulumi logic (can put ur app level logic in `/collab-service/app/` also)
- haven't tested it locally but logic seems sound

zechajw committed:
Fix: minor issues with previous code addition

zechajw committed:
Fix bug where difficulty was not passed correctly

raynerljm committed:
refactor: refactor user card

raynerljm committed:
feat: style dropdown

zechajw committed:
Add: ability to join room given a roomId

- [X] Added different language support to dropdown
- [X] Added /:roomId parameter to room route
- [X] Added /constants/languages.ts for language support

zechajw committed:
Refactor socket

zechajw committed:
Refactor

zechajw committed:
Add integration to pulumi

zechajw committed:
Set pulumi stack

zechajw committed:
Test: github actions
